distributed/shuffle/multi_comm.py

```diff
@@ -0,0 +1,204 @@
+
+        We keep ``max_connections`` comms running while we still have any data
+        as an old comm finishes, we find the next largest buffer, pull off
+        ``max_message_size`` data from it, and ship it to the target worker.
+
+        We do this until we're done.  This coroutine runs in the background.
+
+        See Also
+        --------
+        process: does the actual writing
+        """
+
+        while not self._done:
+            with self.time("idle"):
+                if not self.shards:
+                    await asyncio.sleep(0.1)
+                    continue
+
+                await self.queue.get()
+
+            with self.lock:
+                address = max(self.sizes, key=self.sizes.get)
+
+                size = 0
+                shards = []
+                while size < self.max_message_size:
+                    try:
+                        shard = self.shards[address].pop()
+                        shards.append(shard)
+                        s = len(shard)
+                        size += s
+                        self.sizes[address] -= s
+                    except IndexError:
+                        break
+                    finally:
+                        if not self.shards[address]:
+                            del self.shards[address]
+                            assert not self.sizes[address]
+                            del self.sizes[address]
+
+                assert set(self.sizes) == set(self.shards)
+                assert shards
+                future = asyncio.ensure_future(self.process(address, shards, size))
+                del shards
+                self._futures.add(future)
+
+    async def process(self, address: str, shards: list, size: int):
+        """Send one message off to a neighboring worker"""
+        with log_errors():
+
+            # Consider boosting total_size a bit here to account for duplication
+
+            try:
+                # while (time.time() // 5 % 4) == 0:
+                #     await asyncio.sleep(0.1)
+                start = time.time()
+                with self.time("send"):
+                    await self.send(address, [b"".join(shards)])
+                stop = time.time()
+                self.diagnostics["avg_size"] = (
+                    0.95 * self.diagnostics["avg_size"] + 0.05 * size
+                )
+                self.diagnostics["avg_duration"] = 0.98 * self.diagnostics[
+                    "avg_duration"
+                ] + 0.02 * (stop - start)
+            finally:
+                self.total_size -= size
+                MultiComm.total_size -= size
+                with self.thread_condition:
+                    self.thread_condition.notify()
+                await self.queue.put(None)
+
+    async def flush(self):
+        """
+        We don't expect any more data, wait until everything is flushed through
+        """
+        while self.shards:
+            await asyncio.sleep(0.05)
```

Comment:
self.communicate may create more futures while there are still more shards.  We need to wait until that is done.

Reply:
I would've expected it to be sufficient to first await `_communicate_future` then. 
I tried but this is not possible since the `communicate` never completes if we do not set `_done`. However, if we set it too soon, we will not flush all shards. This feels a bit brittle and I was wondering if we should distinguish "done because we flushed" and "done because an exception was raised", with something like

```python
def communicate(self):
    while not self._exception:
        if self._flushed and not self.shards:
            break
        if not self.shards:
            await asyncio.sleep(0.1)
```

I also receive a bunch of warnings when executing tests about pending tasks

```
Task was destroyed but it is pending!
task: <Task pending name='Task-3542' coro=<MultiComm.communicate() running at /Users/fjetter/workspace/distributed-main/distributed/shuffle/multi_comm.py:129> wait_for=<Future pending cb=[Task.task_wakeup()]>>
```

(same for the multi_file). The line points to the sleep / polling